Index out of range : HyperV collectVmCpuUsage

Good day all,

When running the exporter program with the Hyper V module enabled, it crashes when trying to run the query. I initially thought the issue might be related to the one [here](https://github.com/prometheus-community/windows_exporter/issues/192) because it was the only hyper v host that had a legacy network adapter. Below is the output from the log. Any ideas would be much appreciated. Thanks!

C:\Program Files\windows_exporter>windows_exporter.exe --telemetry.addr=":9184" --collectors.enabled cpu,cs,hyperv,logical_disk,net,os,system --scrape.timeout-margin=2.0 --log.level="debug"
time="2020-10-05T16:33:41-04:00" level=debug msg="Initializing SWbemServices" source="exporter.go:260"
time="2020-10-05T16:33:41-04:00" level=debug msg="Detected Windows version 6.300000\n" source="collector.go:47"
time="2020-10-05T16:33:41-04:00" level=info msg="Enabled collectors: os, system, cpu, cs, hyperv, logical_disk, net" source="exporter.go:338"
time="2020-10-05T16:33:41-04:00" level=info msg="Starting windows_exporter (version=0.14.0, branch=master, revision=922c08b85b60da8a444a57a84126448d7c993307)" source="exporter.go:378"
time="2020-10-05T16:33:41-04:00" level=info msg="Build context (go=go1.13.3, user=appvyr-win\\appveyor@APPVYR-WIN, date=20200907-07:33:36)" source="exporter.go:379"
time="2020-10-05T16:33:41-04:00" level=info msg="Starting server on :9184" source="exporter.go:382"
time="2020-10-05T16:33:56-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_ComputerSystem" source="wmi.go:24"
time="2020-10-05T16:33:56-04:00" level=debug msg="collector logical_disk succeeded after 0.001005s." source="exporter.go:221"
time="2020-10-05T16:33:56-04:00" level=debug msg="collector net succeeded after 0.001005s." source="exporter.go:221"
time="2020-10-05T16:33:56-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_OperatingSystem" source="wmi.go:24"
time="2020-10-05T16:33:56-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_PerfRawData_VmmsVirtualMachineStats_HyperVVirtualMachineHealthSummary" source="wmi.go:24"
time="2020-10-05T16:33:56-04:00" level=debug msg="collector system succeeded after 0.000000s." source="exporter.go:221"
time="2020-10-05T16:33:56-04:00" level=debug msg="collector cpu succeeded after 0.001998s." source="exporter.go:221"
time="2020-10-05T16:33:56-04:00" level=debug msg="collector cs succeeded after 0.028817s." source="exporter.go:221"
time="2020-10-05T16:33:56-04:00" level=debug msg="collector os succeeded after 0.068321s." source="exporter.go:221"
time="2020-10-05T16:33:56-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_PerfRawData_VidPerfProvider_HyperVVMVidPartition" source="wmi.go:24"
time="2020-10-05T16:33:56-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_PerfRawData_HvStats_HyperVHypervisorRootPartition" source="wmi.go:24"
time="2020-10-05T16:33:57-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_PerfRawData_HvStats_HyperVHypervisor" source="wmi.go:24"
time="2020-10-05T16:33:57-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_PerfRawData_HvStats_HyperVHypervisorRootVirtualProcessor" source="wmi.go:24"
time="2020-10-05T16:33:57-04:00" level=debug msg="Generated WMI query SELECT * FROM Win32_PerfRawData_HvStats_HyperVHypervisorVirtualProcessor" source="wmi.go:24"
panic: runtime error: index out of range [2] with length 1

goroutine 199 [running]:
github.com/prometheus-community/windows_exporter/collector.(*HyperVCollector).collectVmCpuUsage(0xc0000d2f00, 0xc0000d5ce0, 0x0, 0x0, 0x0)
        C:/gopath/src/github.com/prometheus-community/windows_exporter/collector/hyperv.go:1005 +0x992
github.com/prometheus-community/windows_exporter/collector.(*HyperVCollector).Collect(0xc0000d2f00, 0xc0000060b8, 0xc0000d5ce0, 0x0, 0x0)
        C:/gopath/src/github.com/prometheus-community/windows_exporter/collector/hyperv.go:626 +0x56d
main.execute(0xc00001a217, 0x6, 0xb801a0, 0xc0000d2f00, 0xc0000060b8, 0xc0000d5ce0, 0x0)
        C:/gopath/src/github.com/prometheus-community/windows_exporter/exporter.go:208 +0x8f
main.windowsCollector.Collect.func2(0xc0002bfb30, 0xc0000060b8, 0xc0000d5ce0, 0xc0002bfb40, 0xc0002bfb48, 0xc0001d65a0, 0xc00001a217, 0x6, 0xb801a0, 0xc0000d2f00)
        C:/gopath/src/github.com/prometheus-community/windows_exporter/exporter.go:149 +0xb1
created by main.windowsCollector.Collect
        C:/gopath/src/github.com/prometheus-community/windows_exporter/exporter.go:147 +0x671

--- a/collector/hyperv.go
+++ b/collector/hyperv.go
@@ -1001,8 +1001,17 @@ func (c *HyperVCollector) collectVmCpuUsage(ch chan<- prometheus.Metric) (*prome
 		}
 		// The name format is <VM Name>:Hv VP <vcore id>
 		parts := strings.Split(obj.Name, ":")
+		if len(parts) != 2 {
+			log.Warnf("Unexpected format of Name in collectVmCpuUsage: %q, expected %q. Skipping.", obj.Name, "<VM Name>:Hv VP <vcore id>")
+			continue
+		}
+		coreParts := strings.Split(parts[1], " ")
+		if len(coreParts) != 3 {
+			log.Warnf("Unexpected format of core identifier in collectVmCpuUsage: %q, expected %q. Skipping.", parts[1], "Hv VP <vcore id>")
+			continue
+		}
 		vmName := parts[0]
-		coreId := strings.Split(parts[1], " ")[2]
+		coreId := coreParts[2]
 
 		ch <- prometheus.MustNewConstMetric(
 			c.VMGuestRunTime,

